lib/common/operations.c

```diff
@@ -44,83 +48,157 @@ pcmk__op_key(const char *rsc_id, const char *op_type, guint interval_ms)
-        *rsc_id = NULL;
-    }
-    if (op_type) {
-        *op_type = NULL;
-    }
-    if (interval_ms) {
-        *interval_ms = 0;
-    }
+    size_t op_len = 0;
 
-    CRM_CHECK(key && *key, return FALSE);
-
-    // Parse interval at end of string
     len = strlen(key);
     offset = len - 1;
+
+    // Parse interval at end of string
     while ((offset > 0) && isdigit(key[offset])) {
-        ch = key[offset] - '0';
-        for (int digits = len - offset; digits > 1; --digits) {
-            ch = ch * 10;
-        }
-        local_interval_ms += ch;
         offset--;
     }
+
     if (interval_ms) {
-        *interval_ms = local_interval_ms;
+        unsigned long long l;
+
+        errno = 0;
+        l = strtoul(&(key[offset+1]), NULL, 10);
+
+        if (errno != 0) {
+            return FALSE;
+        }
+
+        *interval_ms = (guint) l;
     }
 
-    CRM_CHECK((offset != (len - 1)) && (key[offset] == '_'), return FALSE);
+    // Verify we're at the separator between the operation and interval.
+    if (offset == len-1 || key[offset] != '_') {
+        if (interval_ms) {
+            *interval_ms = 0;
+        }
 
-    mutable_key = strndup(key, offset);
+        return FALSE;
+    }
+
+    // We're already on the underscore before the interval.  Back up one
+    // or this loop will never do anything.
     offset--;
 
-    while (offset > 0 && key[offset] != '_') {
+    while ((offset > 0) && key[offset] != '_') {
         offset--;
+        op_len++;
     }
 
-    CRM_CHECK(key[offset] == '_',
-              free(mutable_key); return FALSE);
+    if (op_type) {
+        // Add one here to skip the leading underscore we landed on in the
+        // while loop.
+        *op_type = strndup(&(key[offset+1]), op_len);
+    }
 
-    mutable_key_ptr = mutable_key + offset + 1;
+    // Verify we're at the separator between the resource and operation.
+    if (offset == len-1 || key[offset] != '_') {
+        if (interval_ms) {
+            *interval_ms = 0;
+        }
 
-    if (op_type) {
-        *op_type = strdup(mutable_key_ptr);
+        if (op_type) {
+            free(*op_type);
+            *op_type = NULL;
+        }
+
+        return FALSE;
     }
 
-    mutable_key[offset] = 0;
-    offset--;
+    // Everything else is the name of the resource.
+    if (rsc_id) {
+        *rsc_id = strndup(key, offset);
+    }
+
+    return TRUE;
+}
 
-    notify = strstr(mutable_key, "_post_notify");
-    if (notify && pcmk__str_eq(notify, "_post_notify", pcmk__str_casei)) {
-        notify[0] = 0;
+static gboolean
+try_migrate_notify_match(const char *key, char **rsc_id, char **op_type, guint *interval_ms)
+{
+    int rc = 0;
+    size_t nmatch = 8;
+    regmatch_t *pmatch = NULL;
+
+    if (notify_migrate_re == NULL) {
+        notify_migrate_re = calloc(1, sizeof(regex_t));
+        rc = regcomp(notify_migrate_re, "^(.*)_(migrate_(from|to)|(pre|post)_notify_([a-z]+|migrate_(from|to)))_([0-9]+)$",
+                     REG_EXTENDED);
```

Comment:
Nothing to worry about now, but valgrind is probably going to complain about this. We may need to add a suppression for it. Eventually, I'd like to have a cleanup function for the library that would free stray memory like this, but it's not a priority at this point.

Reply:
Okay, I think I've done that in the right place and with the right syntax.  I've pushed it so feel free to give it a double check.

Would it be possible to use the gcc `__attribute__((destructor))` stuff to clean up on shut down?  I don't know if that's supported in other compilers or if we especially care about that.  Still, it might be a handy thing to use.  I'm all for using whatever the compiler gives us.

Reply:
That should work for cppcheck, but valgrind has cts/valgrind-pcmk.suppressions. You can add that if necessary when valgrinding 2.0.5.

We'd want to functionize the clean-up and call it on platforms without destructor, so we might as well just do that everywhere. The drudgework will be going through all the code looking for globals, and since the only real advantage is to make valgrind quieter, I'm not too concerned right now. What I kind of imagine is: ideally all globals would be static to one file, with accessor functions as needed; each file with globals would define its own clean-up function; each library would define a library clean-up function that would call of the file clean-up functions. Also avoiding globals is even better, where we can easily move them into an object with new/free API.